src/main/java/com/amazon/opendistroforelasticsearch/sql/parser/subquery/InSubqueryRewriter.java

```diff
@@ -0,0 +1,146 @@
+    private final SQLInSubQueryExpr inSubQueryExpr;
+    private final MySqlSelectQueryBlock inSubQuery;
+    private final SQLExpr joinLeftExpr;
+    private final SQLExpr joinRightExpr;
+
+    public InSubqueryRewriter(Subquery subquery) {
+        this.inSubQueryExpr = (SQLInSubQueryExpr) subquery.getSubQueryExpr();
+        this.inSubQuery = (MySqlSelectQueryBlock) inSubQueryExpr.getSubQuery().getQuery();
+        joinLeftExpr = inSubQueryExpr.getExpr();
+        joinRightExpr = rightExpr(inSubQuery);
+    }
+
+    /**
+     * Rewrite the subquery with JOIN clause. This API is NOT idempotent.
+     *
+     * @param query Subquery will be written.
+     */
+    @Override
+    public void rewrite(MySqlSelectQueryBlock query) {
+        query.setFrom(buildFrom(query));
+        query.setWhere(buildWhere(query));
+    }
+
+    /**
+     * Build From clause from input query.
+     *
+     * With the input query.
+     *       Query
+     *     /   |   \
+     * SELECT FROM     WHERE
+     *    |    |     /   |  \
+     *    *    A  c>10 AND INSubquery
+     *                      /    \
+     *                     a    Query
+     *                          /    \
+     *                       SELECT FORM
+     *                         |     |
+     *                         b     B
+     *
+     * The FROM logic should be
+     *     JOIN
+     *     / | \
+     *    A  B A.a == B.b
+     */
+    private SQLJoinTableSource buildFrom(MySqlSelectQueryBlock rootQuery) {
+        SQLJoinTableSource sqlJoin = new SQLJoinTableSource();
+        sqlJoin.setLeft(rootQuery.getFrom());
+        sqlJoin.setRight(inSubQuery.getFrom());
+        sqlJoin.setCondition(new SQLBinaryOpExpr(joinLeftExpr, SQLBinaryOperator.Equality, joinRightExpr));
+        sqlJoin.setJoinType(JoinType.JOIN);
+
+        return sqlJoin;
+    }
+
+    /**
+     * Build Where clause from input query.
+     *
+     * With the input query.
+     *       Query
+     *     /   |   \
+     * SELECT FROM     WHERE
+     *    |    |     /   |  \
+     *    *    A  c>10 AND INSubquery
+     *                      /    \
+     *                     a    Query
+     *                          /    \
+     *                       SELECT FORM
+     *                         |     |
+     *                         b     B
+     *
+     *
+     *     WHERE
+     *     /  |   \
+     *   c>10 AND B.b is NOT NULL
+     */
+    private SQLExpr buildWhere(MySqlSelectQueryBlock rootQuery) {
+        return rewriteSubquery(rootQuery.getWhere());
+    }
+
+    private SQLExpr rewriteSubquery(SQLExpr expr) {
+        if (expr instanceof SQLInSubQueryExpr) {
+            return createSubqueryReplacementCondition();
+        }
+        else if (expr instanceof SQLBinaryOpExpr) {
+            ((SQLBinaryOpExpr) expr).setLeft(rewriteSubquery(((SQLBinaryOpExpr) expr).getLeft()));
+            ((SQLBinaryOpExpr) expr).setRight(rewriteSubquery(((SQLBinaryOpExpr) expr).getRight()));
+            return expr;
```

Comment:
Can you `return expr;` out of if-else blocks and only once?

Reply:
I support existing code as it is. 
Only one exit makes code harder to read
https://blog.codinghorror.com/flattening-arrow-code/

> Always opportunistically return as soon as possible from the function. Once your work is done, get the heck out of there! This isn't always possible – you might have resources you need to clean up. But whatever you do, you have to abandon the ill-conceived idea that there should only be one exit point at the bottom of the function.

Reply:
@galkk  I wouldn't take someone's recommendation as a hard rule to follow in every case. I prefer avoiding duplicated code, if you need to return a value, but conditionally do something before that, why return that value twice, just do whatever you want to do conditionally and return that value. The only thing that makes me agree to current code is that there is another condition above (the `if` block) that I somehow missed.